File: frontend/src/lib/api/credit.ts
// /api/v1/credit 客户端
//
// 后端契约见 backend/app/api/v1/credit.py
// 技术方案 §四

import { api } from "../api";

export type Grade = "A" | "B" | "C" | "D";

export interface CompanyListItem {
  id: number;
  name: string;
  legal_name_en: string | null;
  country_code: string;
  registration_no: string | null;
  total_score: number | null;
  grade: Grade | null;
}

export interface DimensionOut {
  code: string;
  name: string;
  max_score: number;
  score: number;
}

export interface ScoreDetailOut {
  dimension_code: string;
  dimension_name: string;
  subitem_code: string;
  subitem_name: string;
  score: number;
  max_score: number;
  hit_rule_code: string | null;
  hit_rule_description: string | null;
  is_default_score: boolean;
}

export interface SnapshotOut {
  id: number;
  total_score: number;
  grade: Grade;
  rule_version: number;
  trigger_type: string;
  ai_summary: string | null;
  ai_summary_generated_at: string | null;
  is_current: boolean;
  calculated_at: string;
}

export interface BasicDataOut {
  established_date: string | null;
  registered_capital: string | null;
  business_scope: string | null;
  legal_representative: string | null;
  shareholders: string | null;
  status_text: string | null;
  address: string | null;
  website: string | null;
  data_source: string;
  fetched_at: string | null;
}

export interface FinanceDataOut {
  revenue_trend: string | null;
  debt_ratio: string | null; // Decimal 序列化为字符串
  cash_flow_status: string | null;
  data_source: string;
  fetched_at: string | null;
}

export interface LegalDataOut {
  litigation_count: number;
  defaulter_unresolved_count: number;
  defaulter_resolved_count: number;
  negative_news_level: string | null;
  data_source: string;
  fetched_at: string | null;
}

export interface CertificationOut {
  id: number;
  cert_type: "mandatory_country" | "system_general" | "industry_specific";
  cert_name: string;
  target_country_code: string | null;
  issuer: string | null;
  issued_at: string | null;
  expires_at: string | null;
  status: "valid" | "expired" | "suspicious_forged";
  data_source: string;
}

export interface CompanyDetailOut {
  id: number;
  name: string;
  legal_name_en: string | null;
  country_code: string;
  registration_no: string | null;
  snapshot: SnapshotOut | null;
  dimensions: DimensionOut[];
  details: ScoreDetailOut[];
  basic: BasicDataOut | null;
  finance: FinanceDataOut | null;
  legal: LegalDataOut | null;
  certifications: CertificationOut[];
}

export interface SearchHistoryItem {
  id: number;
  company_id: number;
  company_name: string;
  country_code: string;
  grade: Grade | null;
  searched_at: string;
}

export interface AiMessage {
  id: number;
  role: "user" | "assistant";
  content: string;
  sequence: number;
  created_at: string;
}

export interface AiConversation {
  id: number;
  user_id: number;
  company_id: number;
  started_at: string;
  messages: AiMessage[];
}

export const creditApi = {
  /** 搜索候选企业。 */
  search: (params: { country?: string; q?: string }) => {
    const qs = new URLSearchParams();
    if (params.country) qs.set("country", params.country);
    if (params.q) qs.set("q", params.q);
    return api.get<CompanyListItem[]>(
      `/api/v1/credit/companies/search?${qs.toString()}`
    );
  },

  /** 企业详情(首访可能触发 ai_summary 同步生成,响应可能稍慢)。 */
  detail: (companyId: number) =>
    api.get<CompanyDetailOut>(`/api/v1/credit/companies/${companyId}`),

  /** 单家重算(OPERATOR / ADMIN)。 */
  recompute: (companyId: number) =>
    api.post<SnapshotOut>(`/api/v1/credit/companies/${companyId}/recompute`),

  /** 当前用户最近 5 条搜索历史。 */
  history: () => api.get<SearchHistoryItem[]>("/api/v1/credit/search-history"),

  /** 删除一条历史。 */
  deleteHistory: (id: number) =>
    api.delete<{ deleted: number }>(`/api/v1/credit/search-history/${id}`),

  /** 创建 AI 会话。 */
  createConversation: (companyId: number) =>
    api.post<AiConversation>("/api/v1/credit/ai/conversations", {
      company_id: companyId,
    }),

  /** 获取会话历史。 */
  getConversation: (id: number) =>
    api.get<AiConversation>(`/api/v1/credit/ai/conversations/${id}`),
};

/** SSE 流式发送消息。每收到一段 chunk 调 onChunk;[DONE] 或 error 时调 onDone。
 *
 * 直接 fetch 走 ReadableStream(避免 EventSource 不能带 Authorization 头)。
 */
export async function streamAiMessage(
  conversationId: number,
  content: string,
  callbacks: {
    onChunk?: (chunk: string) => void;
    onDone?: () => void;
    onError?: (msg: string) => void;
  }
): Promise<void> {
  const { useAuthStore } = await import("@/stores/authStore");
  const accessToken = useAuthStore.getState().accessToken;
  const BASE =
    process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:8000";
  const resp = await fetch(
    `${BASE}/api/v1/credit/ai/conversations/${conversationId}/messages`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "text/event-stream",
        ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
      },
      credentials: "include",
      body: JSON.stringify({ content }),
    }
  );

  if (!resp.ok) {
    callbacks.onError?.(`HTTP ${resp.status}`);
    return;
  }
  const reader = resp.body?.getReader();
  if (!reader) {
    callbacks.onError?.("无响应流");
    return;
  }
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    // SSE 帧间用 \n\n 分隔
    let idx: number;
    while ((idx = buffer.indexOf("\n\n")) !== -1) {
      const frame = buffer.slice(0, idx);
      buffer = buffer.slice(idx + 2);
      // 解析帧:可能是 "event: error\ndata: ..." 或 "data: ..."
      const lines = frame.split("\n");
      let evt = "message";
      let data = "";
      for (const line of lines) {
        if (line.startsWith("event: ")) evt = line.slice(7);
        else if (line.startsWith("data: ")) data = line.slice(6);
      }
      if (evt === "error") {
        try {
          const parsed = JSON.parse(data);
          callbacks.onError?.(parsed.error || "未知错误");
        } catch {
          callbacks.onError?.(data || "未知错误");
        }
        callbacks.onDone?.();
        return;
      }
      if (data === "[DONE]") {
        callbacks.onDone?.();
        return;
      }
      // 反转义 \n
      callbacks.onChunk?.(data.replace(/\\n/g, "\n"));
    }
  }
  callbacks.onDone?.();
}

File: backend/app/schemas/credit.py
"""信用评估接口 schema(对齐 docs/architecture/信用评估模块技术方案设计-v0_1.md §四 + §六)。"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# =============================================================================
# 候选企业(列表 / 详情共用)
# =============================================================================

class CompanyListItem(BaseModel):
    """搜索结果列表 item。"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    legal_name_en: str | None = None
    country_code: str
    registration_no: str | None = None
    total_score: int | None = None
    grade: str | None = None


class BasicDataOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    established_date: date | None = None
    registered_capital: str | None = None
    business_scope: str | None = None
    legal_representative: str | None = None
    shareholders: str | None = None
    status_text: str | None = None
    address: str | None = None
    website: str | None = None
    data_source: str
    fetched_at: datetime | None = None


class FinanceDataOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    revenue_trend: str | None = None
    debt_ratio: Decimal | None = None
    cash_flow_status: str | None = None
    data_source: str
    fetched_at: datetime | None = None


class LegalDataOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    litigation_count: int
    defaulter_unresolved_count: int
    defaulter_resolved_count: int
    negative_news_level: str | None = None
    data_source: str
    fetched_at: datetime | None = None


class CertificationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    cert_type: str
    cert_name: str
    target_country_code: str | None = None
    issuer: str | None = None
    issued_at: date | None = None
    expires_at: date | None = None
    status: str
    data_source: str


class ScoreDetailOut(BaseModel):
    """单子项明细。"""
    model_config = ConfigDict(from_attributes=True)
    dimension_code: str
    dimension_name: str
    subitem_code: str
    subitem_name: str
    score: int
    max_score: int
    hit_rule_code: str | None = None
    hit_rule_description: str | None = None
    is_default_score: bool


class DimensionOut(BaseModel):
    """维度元信息 + 当前分(用于前端雷达图)。"""
    code: str
    name: str
    max_score: int
    score: int


class DimensionOverrideHitOut(BaseModel):
    """命中的维度级 override 明细(v0.2)。"""
    dimension_code: str
    override_rule_code: str
    override_description: str
    natural_score: int
    final_score: int


class SnapshotOut(BaseModel):
    """评分快照。"""
    model_config = ConfigDict(from_attributes=True)
    id: int
    total_score: int
    grade: str
    # v0.2:同时返回自然分(未 override)和最终分(各维度 dimension_N_score)
    # 历史 v0.1 快照这 5 字段为 NULL,前端等价 '无 override'
    dimension_1_natural_score: int | None = None
    dimension_2_natural_score: int | None = None
    dimension_3_natural_score: int | None = None
    dimension_4_natural_score: int | None = None
    dimension_overrides: list[DimensionOverrideHitOut] | None = None
    rule_version: int
    trigger_type: str
    ai_summary: str | None = None
    ai_summary_generated_at: datetime | None = None
    is_current: bool
    calculated_at: datetime


class CompanyDetailOut(BaseModel):
    """详情页响应。"""
    id: int
    name: str
    legal_name_en: str | None = None
    country_code: str
    registration_no: str | None = None
    snapshot: SnapshotOut | None = None
    dimensions: list[DimensionOut] = Field(default_factory=list)
    details: list[ScoreDetailOut] = Field(default_factory=list)
    basic: BasicDataOut | None = None
    finance: FinanceDataOut | None = None
    legal: LegalDataOut | None = None
    certifications: list[CertificationOut] = Field(default_factory=list)


# =============================================================================
# 搜索历史
# =============================================================================

class SearchHistoryItem(BaseModel):
    id: int
    company_id: int
    company_name: str
    country_code: str
    grade: str | None = None
    searched_at: datetime


# =============================================================================
# AI 会话
# =============================================================================

class AiConversationCreateIn(BaseModel):
    company_id: int


class AiMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    role: str
    content: str
    sequence: int
    created_at: datetime


class AiConversationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    user_id: int
    company_id: int
    started_at: datetime
    messages: list[AiMessageOut] = Field(default_factory=list)


class AiMessageSendIn(BaseModel):
    content: str = Field(..., min_length=1, max_length=2000)


# =============================================================================
# 触发详情(写入 score_snapshot.trigger_detail)
# =============================================================================

class TriggerDetail(BaseModel):
    """评分触发上下文(放进 trigger_detail JSONB)。"""
    actor_user_id: int | None = None
    actor_email: str | None = None
    source: str | None = None  # api / seed / batch / event
    extra: dict[str, Any] | None = None
